src/index.ts

```diff
@@ -1,9 +1,230 @@
+        }
+
+        return token[1] + value;
+    });
+
+    return compiled.reverse().join("");
+}
+
+/**
+ * Stringifies a value for usage as path / query / host parameter or fragment
+ */
+function stringifyValue (value: string|number|boolean|null|undefined, explicitFalse: boolean = false) : string
+{
+    if (true === value)
+    {
+        return "1";
+    }
+
+    if (false === value)
+    {
+        return explicitFalse ? "0" : "";
+    }
+
+    return null != value
+        ? value.toString()
+        : "";
+}
+
+/**
+ * Generates the query string
+ */
+function generateQuery (variables: string[], parameters: RouteParameters) : string|null
+{
+    // use all parameters as default
+    let surplus: RouteParameters = parameters;
+
+    // set all parameters to undefined that are already "used" as variables
+    variables.forEach(name => surplus[name] = undefined);
+
+    // filter all null / undefined parameters
+    let filtered: RouteParameters = {};
+    for (let name in surplus)
+    {
+        if (!hasOwnProperty(surplus, name))
+        {
+            continue;
+        }
+
+        if (null != surplus[name] && "_fragment" !== name)
+        {
+            filtered[name] = stringifyValue(surplus[name], true);
+        }
+    }
+
+    return Object.keys(filtered).length > 0
+        ? encode(filtered, "?")
+        : "";
+}
+
+/**
+ * Generates the fragment part
+ */
+function generateFragment (parameters: RouteParameters) : string
+{
+    let fragment = parameters["_fragment"];
+
+    return !fragment
+        ? ""
+        : "#" + stringifyValue(fragment);
+}
+
+/**
+ * Router, that generates URLs
+ */
 export class Router
 {
+    private context: Context = {
+        baseUrl: "",
+        host: "",
+        ports: DEFAULT_PORTS,
+        scheme: "https",
+    };
+    private routes: RouteCollection = {};
+
+
+    /**
+     * Initializes the router
+     */
+    public init (data: BecklynJavaScriptRouter.InitData)
+    {
+        this.context = data.context;
+        this.routes = data.routes;
+    }
+
+
+    /**
+     * Generates a URL or path for a specific route based on the given parameters.
+     */
+    public generate (name: string, parameters: RouteParameters = {}, referenceType: ReferenceType = ReferenceType.ABSOLUTE_PATH) : string
+    {
+        let route = this.routes[name];
+
+        //region Check for missing route
+        if (undefined === route)
+        {
+            throw new Error(`Route not found: '${name}'`);
+        }
+        //endregion
+
+
+        //region Check for missing parameters
+        let missingParameters = route.variables.filter(
+            variable => {
+                return undefined === parameters[variable];
+            }
+        );
+
+        if (missingParameters.length > 0)
+        {
+            throw new Error(`Some mandatory parameters are missing ("${missingParameters.join('", "')}")`);
```

Comment:
We should add the route's name to improve debugging experience.

```suggestion
            throw new Error(`Some mandatory parameters for route '${route}' are missing ("${missingParameters.join('", "')}")`);
```

Reply:
Ah yes, previously the check was in a nested function, so it would have been catched by the global catch. Now we need to add it ourselves.